File: cpp/LeetCode/No84.cpp
/*
Given an array of integers heights representing the histogram's bar height 
where the width of each bar is 1, return the area of the largest rectangle in the histogram.
*/
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack<int> s;
        int max_area = 0;
        for(int i=0;i<heights.size();i++){
            while(s.size()>0 && heights[i]<heights[s.top()]){
                int index = s.top();
                s.pop();
                int h = heights[index];
                if(s.size()>0){
                    int area = h*(i-s.top()-1);
                    max_area = max(max_area, area);
                }
                else{
                    int area = h*(i);
                    max_area = max(max_area, area);
                }
                
            }
            s.push(i);
        }

        while(s.size()>0){
            int index = s.top();
            s.pop();
            int h = heights[index];
            
            if(s.size()>0){
                int area = h*(heights.size()-s.top()-1);
                max_area = max(max_area, area);
            }
            else{
                int area = h*(heights.size());
                max_area = max(max_area, area);
            }
        }
        return max_area;
    }
};

int main(){
    Solution solution;
    
    vector<int> heights{5,4,1,2};
    cout << solution.largestRectangleArea(heights) << endl;
    return 0;
}

File: python/LeetCode Quiz/No84.py
'''
Given an array of integers heights representing the histogram's bar height where the width of each bar is 1
, return the area of the largest rectangle in the histogram.
'''
from typing import List

class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        max_area = 0
        for i, h in enumerate(heights):
            while len(stack)>0 and h<heights[stack[-1]]:
                height = heights[stack.pop()]
                if len(stack)>0:
                    max_area = max(max_area, height*(i-stack[-1]-1))
                else:
                    max_area = max(max_area, height*i)
            stack.append(i)
        while len(stack)>0:
            height = heights[stack.pop()]
            if len(stack)>0:
                max_area = max(max_area, height*(len(heights)-stack[-1]-1))
            else:
                max_area = max(max_area, height*len(heights))
        return max_area

def main():
    solution = Solution()

    heights = [2,1,5,6,2,3]
    print(solution.largestRectangleArea(heights))

if __name__=='__main__':
    main()
